src/main/java/htsjdk/samtools/util/BlockCompressedIndex.java

```diff
@@ -0,0 +1,248 @@
+     * @return the number of blocks.
+     */
+    public int getNumberOfBlocks() {
+        return entries.length;
+    }
+
+    /**
+     * Gets an unmodifiable list with the index entries.
+     *
+     * @return index entries.
+     */
+    public List<IndexEntry> getIndexEntries() {
+        return Collections.unmodifiableList(Arrays.asList(entries));
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        if (obj == null || !(obj instanceof BlockCompressedIndex)) {
+            return false;
+        }
+        return Arrays.equals(this.entries, ((BlockCompressedIndex) obj).entries);
+    }
+
+    @Override
+    public int hashCode() {
+        return Arrays.hashCode(entries);
+    }
+
+    @Override
+    public String toString() {
+        return String.format("BlockCompressedIndex:%s", Arrays.toString(entries));
+    }
+
+    /**
+     * Writes this index into the requested path.
+     *
+     * @param output the output file.
+     *
+     * @throws IOException if an I/O error occurs.
+     */
+    public void writeIndex(final Path output) throws IOException {
+        if (output == null) {
+            throw new IllegalArgumentException("null output path");
+        }
+
+        final ByteBuffer buffer = allocateBuffer(entries.length, true);
+
+        // put the number of entries
+        buffer.putLong(Integer.toUnsignedLong(entries.length));
+
+        for (final IndexEntry entry : entries) {
+            // TODO - check if it is unsigned?
+            buffer.putLong(entry.getCompressedOffset());
+            buffer.putLong(entry.getUncompressedOffset());
+        }
+
+        // write into the output
+        Files.write(output, buffer.array());
+    }
+
+    /**
+     * Load the index from the provided file.
+     *
+     * @param indexPath the path for the index to load.
+     *
+     * @return loaded index.
+     *
+     * @throws IOException if an I/O error occurs.
+     */
+    public static final BlockCompressedIndex loadIndex(final Path indexPath) throws IOException {
+        // allocate a buffer for re-use for read each byte
+        ByteBuffer buffer = ByteBuffer.allocate(Long.BYTES);
+        buffer.order(ByteOrder.LITTLE_ENDIAN);
+
+        try (final ByteChannel channel = Files.newByteChannel(indexPath)) {
+            if (Long.BYTES != channel.read(buffer)) {
+                throw new IOException("corrupted index file");
+            }
+            buffer.flip();
+            final int numberOfEntries = (int) buffer.getLong();
+
+            // allocate array with the entries
+            final IndexEntry[] entries = new IndexEntry[numberOfEntries];
+
+            // create a new buffer with the correct size and read into it
+            buffer = allocateBuffer(numberOfEntries, false);
+            channel.read(buffer);
+            buffer.flip();
+
+            for (int i = 0; i < numberOfEntries; i++) {
+                entries[i] = new IndexEntry(buffer.getLong(), (int) buffer.getLong());
```

Comment:
Could you add checks here that the index entries values are increasing? It would be an inexpensive check that would catch a lot of potential index corruption.  You might want to check that they're positive as well.

Reply:
That's a very good indea. I'll add in next commit.

Reply:
Done.